Todo quickfix

fancsali opened the pull request:
Steps for submitting a pull request:

- [X]  **ALL** pull requests should be made against the `dev` branch!
- [ ] Take a look at [CONTRIBUTING.MD](https://github.com/vimwiki/vimwiki/blob/dev/CONTRIBUTING.md)
- [X] Reference any related issues.
    - This is intended to address the issue of collecting **todo items**, described in #515 (amongst [tips in the wiki](https://vimwiki.github.io/vimwikiwiki/Tips%20and%20Snips.html#Tips%20and%20Snips-Task%20Management-Find%20Incomplete%20Tasks) and discussions here on GitHub and elsewhere online)
- [X] Provide a description of the proposed changes.
    - This will collect all **todo items** from the **diary** and present them in a quickfix buffer, sorted from oldest to newest; allowing the user to toggle the items from the list with `<C-Space>` directly from the list. 
- [ ] PRs must pass Vint tests and add new Vader tests as applicable.
- [ ] Make sure to update the documentation in `doc/vimwiki.txt` if applicable,
      including the Changelog and Contributors sections.

Pending/planned before merging:

- [ ] Use **location list** instead of the **quickfix list**
- [ ] Cross check with the [relevant section of the wiki](https://vimwiki.github.io/vimwikiwiki/Tips%20and%20Snips.html#Tips%20and%20Snips-Task%20Management-Find%20Incomplete%20Tasks)
- [ ] Provide mapping (also based on [wiki](https://vimwiki.github.io/vimwikiwiki/Tips%20and%20Snips.html#Tips%20and%20Snips-Task%20Management-Find%20Incomplete%20Tasks))
- [ ] Map `gln` and `glp` in the list - similar to `<C-Space>`
- [ ] Extend logic to tasks from **all pages** (but keep **diary** as default)
- [ ] Handle file name extension and dynamically
- [ ] Handle syntax dynamically for headings
- [ ] Map `[[` and `]]` to jump between _in-progress_ and _not-yet-started_ tasks (also allow jumping between dates?)
- [ ] Handle all list symbols besides `-`

fancsali committed:
Collect todo items in quickfix list

fancsali committed:
Rename to :VimwikiTodoList and add in-progress items

fancsali committed:
More ViM like output - MarkDown only at the moment

fancsali committed:
Allow toggling items directly from quickfix window

fancsali committed:
Display prefix for files in diary_rel_path in quickfix list

brennen commented:
Nice feature!

Couple of thoughts:

  - For getting the current file extension, I think you want something like `vimwiki#vars#get_wikilocal('ext', wiki_nr)`. (But check me on that.)
  - Other task features handle multiple bullet types - I wonder how hard this would be to add.

fancsali commented:
Thanks for sharing your thoughts, @brennen!

Indeed, it's a WIP feature, and it's clearly somewhat geared towards my workflow with the _file extension_ and the _list types_ - added that to the list of _things to check before merging_...

... also, any more thoughts are very much welcome!

fancsali committed:
Make sure all folds are open before toggling

If folding is enabled when the script jumps to the todo in the original
file, folding can interfere with what is toggled and which lines are
used to update the state in the todo list window/buffer.

Added zR to make sure all and any folds are open before doing anything
else in the original file, and thus making sure only the appropriate
line will be copied back to the quickfix buffer.

fancsali committed:
Location instead of quickfix; simplify for performance

fancsali committed:
Commit Makefile for local test run
